src/edu/ucsb/nceas/metacat/restservice/D1ResourceHandler.java

```diff
@@ -588,4 +600,63 @@ public static String decode(String s) {
         
         return result;
     }
+    
+    /**
+     * Get the session from the header of the request. The route is disabled by default.
+     * 
+     */
+    protected void getSessionFromHeader() {
+        if (enableSessionFromHeader) {
+            logMetacat.debug("D1ResourceHandler.getSessionFromHeader - In the route to get the session from a http header");
+            //check the shared key between Metacat and the http server:
+            if (proxyKey == null || proxyKey.trim().equals("")) {
+                logMetacat.warn("D1ResourceHandler.getSessionFromHeader - Metacat is not configured to handle the feature passing " +
+                                " the certificate by headers since the proxy key is blank");
+                return;
+            }
+            String proxyKeyFromHttp = (String) request.getHeader("X-Proxy-Key");
+            if (proxyKeyFromHttp == null || proxyKeyFromHttp.trim().equals("")) {
+                logMetacat.warn("D1ResourceHandler.getSessionFromHeader - the value of the header X-Proxy-Key is null or blank. " + 
+                                "So Metacat do NOT trust the request.");
+                return;
+            }
+            if (!proxyKey.equals(proxyKeyFromHttp)) {
+                logMetacat.warn("D1ResourceHandler.getSessionFromHeader - the value of the header X-Proxy-Key does not match the one " + 
+                        " stored in Metacat. So Metacat do NOT trust the request.");
+                return;
+            }
+            
+            String verify = (String) request.getHeader("Ssl-Client-Verify");
+            logMetacat.info("D1ResourceHandler.getSessionFromHeader - the status of the ssl client verification is " + verify);
+            if (verify != null && verify.equalsIgnoreCase("SUCCESS")) {
+                //Metacat only looks up the dn from the header when the ssl client was verified.
+                //We confirmed the client couldn't overwrite the value of the header Ssl-Client-Subject-Dn
+                String dn = (String) request.getHeader("Ssl-Client-Subject-Dn");
+                logMetacat.info("D1ResourceHandler.getSessionFromHeader - the ssl client was verified and the subject from the header is " + dn);
+                if (dn != null) {
+                    Subject subject = new Subject();
+                    subject .setValue(dn);
+                    session = new Session();
+                    session.setSubject(subject);
+                    
+                    SubjectInfo subjectInfo = null;
+                    try {
+                        subjectInfo = D1Client.getCN().getSubjectInfo(subject);
+                    } catch (Exception be) {
+                        logMetacat.warn("D1ResourceHandler.getSessionFromHeader - can not get subject information since " + 
```

Comment:
```suggestion
                        logMetacat.warn("D1ResourceHandler.getSessionFromHeader - can not get subject information for subject" + dn + " since " + 
```
It's helplful to know which subject is failing.